### DISCUSSION TITLE
Different Docker datasource max pagination pages between self-hosted and GitHub App

### DISCUSSION THREAD

**Discussion by NPastorale** [renovatebot/renovate#27173] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.180.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

I have found that Renovate behaves differently when running it as a GitHub App and self-hosted.
The App does not find all the dependency updates that exist, it seems like it just ignores them as far as I can tell.

I have performed reproductions of this, you can find the GitHub App managed repo [here](https://github.com/NPastorale/kubernetes/pulls) and the self-hosted one [here](https://github.com/NPastorale/renovate-test/pulls). The code is an exact copy, down to the very commits, so to the best of my knowledge that shouldn't be the cause.

You'll see that presently there are 3 PRs on the self-hosted version and 1 on the App. The `descheduler` one can be safely ignored as that is appearing consistently. The other two are the ones that differ.

I haven't found why in the logs. [App](https://gist.github.com/NPastorale/c60f9e0bdd0c1d9d97ff1233cf0898f0) [Self-hosted](https://gist.github.com/NPastorale/2d7abf16b48cefd6336044c89c1aab78)

The config.json for the self-hosted version is the following

```json
{
    "platform": "github",
    "token": "REDACTED",
    "repositories": [
        {
            "repository": "npastorale/renovate-test"
        }
    ]
}
```

Is this expected? Maybe the configuration of the GitHub App actually differs from my frugal self-hosted one and the behaviour I'm seeing is actually intended.

Thanks!

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
Please compare the debug logs between the two and start narrowing down on differences

**Comment by NPastorale** [comment]
I have compared them. I wasn't able to figure out why the self hosted gets the newest version while the GitHub App doesn't. Granted, I'm not familiar with Renovate's logs, but the only thing I could see is that for some reason the GH App is not correctly identifying newer versions.

Taking Prowlarr as an example, you can see [here](https://gist.github.com/NPastorale/2d7abf16b48cefd6336044c89c1aab78#file-gistfile1-txt-L689) that it finds the latest one on the self-hosted, whereas in the [GH App](https://gist.github.com/NPastorale/c60f9e0bdd0c1d9d97ff1233cf0898f0#file-gistfile1-txt-L434) there's no mention of even found images.

I'm happy to take another look if you could tell me whether there's something else I should be looking for.

by the way, this is the command I'm running in case it is relevant
```
docker run --rm -e RENOVATE_CONFIG_FILE=/usr/src/app/config.json -e LOG_LEVEL=debug -v "/files/config.json:/usr/src/app/config.json" renovate/renovate:37.180.0
```

**Comment by rarkins** [comment]
I've done a small test here: https://github.com/renovate-reproductions/d1/pull/3

I set versioning to "loose", which should take the highest sorted tag regardless of compatibility.

So I wonder if there's a stale cache in the hosted app

**Comment by NPastorale** [comment]
A stale cache was one of my hypothesis, but as I saw other dependencies being picked up within hours of being released, and these particular ones that are misbehaving being ignored for days, I discarded it. Having said that, I am unaware of how long such a thing would be cached for.

Also, I had to resort to a regex for versioning, as the images from Linuxserver.io have some mixed versioning schemes. [This](https://github.com/NPastorale/kubernetes/blob/main/renovate.json#L15) is the one I used. You can find the [Regex 101](https://regex101.com/r/zmKQtf/1) tests I performed.

Any chance you can confirm the period for which this may be cached for? Just to know when we can safely discard that as the cause

I would assume that versioning is not at fault, as it works for the self-hosted version. But I did create the regex myself, so that's of course a possibility

**Reply by rarkins** [reply]
My impression was that the longest we'd cache the docker datasource (which is used here) is 3-4 hours

**Reply by rarkins** [reply]
Each job will print a "http statistics" message towards the end which should confirm if the server was queried at all. If not then the answer came from cache

**Reply by NPastorale** [reply]
There's only a http statistic for the current image, i.e. [ls44](https://gist.github.com/NPastorale/c60f9e0bdd0c1d9d97ff1233cf0898f0#file-gistfile1-txt-L2288), but nothing for the expected ls51, nor any intermediate for that matter.
The ls45, which should already be considered a new image and trigger a PR, was [released on December 31st 2023](https://github.com/linuxserver/docker-prowlarr/releases/tag/1.11.4.4173-ls45) so I'm assuming this is not a caching issue. What do you think?

**Reply by rarkins** [reply]
By caching issue I was thinking "caching bug" not "caching too long". So unfortunately it's entirely possible that there's a bug preventing newer tags from being fetched. We'll try to debug it at a low level (e.g. direct Redis querying)

**Reply by rarkins** [reply]
@zharinov @nabeelsaabna can we work out what the Redis cache key would be for this package and then get a raw result out of our production cache?

**Reply by NPastorale** [reply]
> By caching issue I was thinking "caching bug" not "caching too long".

I see. I misunderstood.

I'll be waiting for any update. Let me know whether I can be of any help.

And thank you once again!

**Reply by zharinov** [reply]
I wonder how `lscr.io` host [is cacheable](https://github.com/renovatebot/renovate/blob/0df92a65002e92ef1e6226b4cc40f49ae68dc14e/lib/modules/datasource/docker/index.ts#L961-L965) at all, if not by Docker datasource. Anyway, to resolve this, it would be helpful to have minimal reproduction repo, against which we could run both Renovate GitHub App and the local app, as well as instructions of how to configure Renovate for `lscr.io` auth.

**Reply by rarkins** [reply]
@zharinov reproductions in the original description. 


Another possibility could be to lower limits on page fetching in the hosted app vs default

**Reply by NPastorale** [reply]
@zharinov in case it helps, this issue seems to only happen for two specific images, i.e. Prowlarr and Radarr. There are plenty of other images also coming from `lscr.io` that do not exhibit this behaviour and are being consistently, and quickly, detected as new versions. I have compared manifests of the working images and of those that don't and I haven't also found any difference that would look like the cause to me.

**Reply by rarkins** [reply]
Can you tell how many tags each has and if the newest are sorted first or last?

**Reply by zharinov** [reply]
Well, I managed to run https://github.com/renovate-reproductions/d1 clone on both app and locally, and they look identical:
- https://github.com/renovate-testing/test-27173-cache-mismatch-dev/pull/1
- https://github.com/renovate-testing/test-27173-cache-mismatch-prod/pull/1

**Reply by zharinov** [reply]
Here is a big list of release results (more than 10k items):
https://gist.github.com/zharinov/3bd60a254ab065865fef650fe0821759

**Reply by NPastorale** [reply]
I think this is the same as @rarkins was able to obtain. It seems that the bahaviour I'm seeing only happens when using a [regex for version matching](https://github.com/NPastorale/kubernetes/blob/main/renovate.json#L15).  The new version I'm expecting to see today would be `5.2.6.8376-ls203`, which is within the list you shared.

**Reply by rarkins** [reply]
My question above about sorting is very relevant

**Reply by rarkins** [reply]
@nabeelsaabna what limits do we have in docker tag fetching in the hosted app currently?

**Reply by NPastorale** [reply]
The `5.14` version is somewhat up in the list of results, while the first "upgrade" that I can see in that list that matches the regex, would be [5.2.6.8376-ls197](https://gist.github.com/zharinov/3bd60a254ab065865fef650fe0821759#file-res-json-L9561).

Is it possible that the potential candidates list is being truncated and not queries completely after being returned?

**Reply by zharinov** [reply]
Seems like [`5.14` was released very early](https://gist.github.com/zharinov/3bd60a254ab065865fef650fe0821759#file-res-json-L210) at the project timeline, and since then, the convention of version naming has changed. Renovate doesn't know about it, so `5.14` is being chosen as the greatest version compared to any other (according to common Docker versioning scheme).

**Reply by NPastorale** [reply]
@zharinov my understanding is that with the versioning [regex](https://github.com/NPastorale/kubernetes/blob/main/renovate.json#L15) in place it should only consider matching versioning schemes, thus ignoring the `5.14`, even though it would be a logical upgrade from `5.1.3.8246-ls196` provided the regex wasn't there. Am I understanding it correctly?

**Reply by zharinov** [reply]
Regarding pagination, I get `10073` releases total, with `RENOVATE_X_DOCKER_MAX_PAGES` set to `100` and with page limit of `1000`

**Reply by zharinov** [reply]
@NPastorale Regarding regex usage, yes I think it should filter out all undesired versions, if configured properly. Let me check what's going on with your regex in the runtime...

**Reply by rarkins** [reply]
I suspect it's list truncation in the hosted app (newest tags are last)

**Reply by zharinov** [reply]
Just managed to [receive update to `5.2.6.8376-ls203`](https://github.com/renovate-testing/test-27173-cache-mismatch-dev/pull/1) with [this Renovate config](https://github.com/renovate-testing/test-27173-cache-mismatch-dev/blob/main/renovate.json).

**Reply by zharinov** [reply]
But here we go, once I've applied the same config in the GitHub app, existing pull request was autoclosed, like there is no any new version.

**Reply by zharinov** [reply]
With `RENOVATE_X_DOCKER_MAX_PAGES=7`, I've managed to recieve the same outcome (regex filtering + autoclose) like in GitHub app. So yes, looks like it's the truncation caused by pagination limit. Good catch.

**Reply by NPastorale** [reply]
Can I assume this value will eventually be incremented on the GitHub App so I can trust this will work?
Thank you both for your time!

**Reply by nabeelsaabna** [reply]
I can confirm that the only config option in the hosted app that might affect this are:
`"RENOVATE_X_DOCKER_MAX_PAGES": 7.0,` 
and 
```
"cacheTtlOverride": {
      "datasource-docker-hub-tags": 90.0
    }
```

**Reply by zharinov** [reply]
@rarkins Do you think we should just increase max pages, or some other solution?

**Reply by rarkins** [reply]
If we can't increase them for Docker Hub due to rate limiting, we ideally should have a way of making max pages configurable per-host, or specifically for Docker Hub only
